core/include/drivers/ls_i2c.h

```diff
@@ -0,0 +1,155 @@
+ * The module clears this field without generating a STOP signal when the
+ * master loses arbitration.
+ */
+#define I2C_IBCR_MSSL 0x20
+
+/* 0b - Receive 1b - Transmit */
+#define I2C_IBCR_TXRX 0x10
+
+/* Data acknowledge disable
+ * Values written to this field are only used when the I2C module is a receiver,
+ * not a transmitter.
+ * 0b - The module sends an acknowledge signal to the bus at the 9th clock bit
+ * after receiving one byte of data.
+ * 1b - The module does not send an acknowledge-signal response (that is,
+ * acknowledge bit = 1).
+ */
+#define I2C_IBCR_NOACK 0x08
+
+/* Repeat START
+ * If the I2C module is the current bus master, and you program RSTA=1, the I2C
+ * module generates a repeated START condition. This field always reads as a 0.
+ * If you attempt a repeated START at the wrong time, if the bus is owned by
+ * another master the result is loss of arbitration.
+ */
+#define I2C_IBCR_RSTA 0x04
+
+/* DMA enable */
+#define I2C_IBCR_DMAEN 0x02
+
+/* Transfer Complete */
+#define I2C_IBSR_TCF 0x80
+
+/* I2C bus Busy. 0b - Bus is idle, 1b - Bus is busy */
+#define I2C_IBSR_IBB 0x20
+
+/* Arbitration Lost. software must clear this field by writing a one to it. */
+#define I2C_IBSR_IBAL 0x10
+
+/* I2C bus interrupt flag */
+#define I2C_IBSR_IBIF 0x02
+
+/* Received acknowledge
+ * 0b - Acknowledge received
+ * 1b - No acknowledge received
+ */
+#define I2C_IBSR_RXAK 0x01
+
+/* Bus idle interrupt enable */
+#define I2C_IBIC_BIIE 0x80
+
+/* Glitch filter enable */
+#define I2C_IBDBG_GLFLT_EN 0x08
+
+#define I2C_FLAG_WRITE 0x00000000
+#define I2C_FLAG_READ  0x00000001
+
+#define I2C_BUS_TEST_BUSY      true
+#define I2C_BUS_TEST_IDLE      !I2C_BUS_TEST_BUSY
+#define I2C_BUS_TEST_RX_ACK    true
+#define I2C_BUS_NO_TEST_RX_ACK !I2C_BUS_TEST_RX_ACK
+
+#define I2C_NUM_RETRIES	   500
+
+struct i2c_regs {
+	uint8_t ibad;  /* I2c Bus Address Register */
+	uint8_t ibfd;  /* I2c Bus Frequency Divider Register */
+	uint8_t ibcr;  /* I2c Bus Control Register */
+	uint8_t ibsr;  /* I2c Bus Status Register */
+	uint8_t ibdr;  /* I2C Bus Data I/O Register */
+	uint8_t ibic;  /* I2C Bus Interrupt Config Register */
+	uint8_t ibdbg; /* I2C Bus Debug Register */
+};
+
+/*
+ * sorted list of clock divisor, ibfd register value pairs
+ */
+struct i2c_clock_divisor_pair {
+	uint16_t divisor;
+	uint8_t ibfd; /* I2c Bus Frequency Divider Register value */
+};
+
+/*
+ * I2C device operation
+ * The EFI_I2C_OPERATION describes a subset of an I2C transaction in which
+ * the I2C controller is either sending or receiving bytes from the bus.
+ * Some transactions will consist of a single operation while others will
+ * be two or more.
+ */
+struct i2c_operation {
+	/* Flags to qualify the I2C operation. */
+	uint32_t flags;
+
+	/*
+	 * Number of bytes to send to or receive from the I2C device.  A ping
+	 * (address only byte/bytes)  is indicated by setting the LengthInBytes
+	 * to zero.
+	 */
+	uint32_t length_in_bytes;
```

Comment:
Expected type here would be `unsigned int` or `size_t` depending on how large the buffer might be.